hil/ext/switches/juniper.py

```diff
@@ -0,0 +1,434 @@
+        all_output = []
+        for p_obj in ports:
+            port = p_obj.label
+            port_info = self._interface_info(port)
+            native_no = port_info[port]['native_vlan']
+            vlans = port_info[port]['vlans']
+            if vlans == 'default':
+                response[p_obj] = []
+            elif vlans == native_no:
+                response[p_obj] = [('vlan/native', str(native_no))]
+            elif native_no is None and isinstance(vlans, (str, unicode)):
+                response[p_obj] = [('vlan/'+str(vlans), str(vlans))]
+            elif native_no is None and isinstance(vlans, list):
+                for vlan in vlans:
+                    all_output.append(('vlan/'+str(vlan), str(vlan)))
+                response[p_obj] = filter(
+                        lambda x: x[0] not in [
+                            'vlan/default', 'vlan/'+str(native_no)
+                            ], all_output
+                        )
+            else:
+                native_no = str(native_no)
+                all_output = [('vlan/native', native_no)]
+                for vlan in port_info[port]['vlans']:
+                    all_output.append(('vlan/'+str(vlan), str(vlan)))
+                response[p_obj] = filter(
+                        lambda x: x[0] not in [
+                            'vlan/default', 'vlan/'+str(native_no)
+                            ], all_output
+                        )
+        return response
+
+    def _set_native_vlan(self, port, network_id):
+        """Sets native vlan for a trunked port.
+        It enables the port, if it is the first vlan for the port.
+        """
+        port_info = self._interface_info(port)
+        self._set_commands(port, network_id)
+        set_native_vlan = self.set_native_vlan
+        if not port_info[port]['trunk_mode']:
+            set_native_vlan = self.set_trunk_mode + set_native_vlan
+
+        if port_info[port]['disabled']:
+            set_native_vlan = self.enable_interface + set_native_vlan
+
+        if 'default' in port_info[port]['vlans']:
+            set_native_vlan = set_native_vlan + self.remove_default_vlan
+
+        self._set_commit_config(set_native_vlan)
+        self._clear_set_commands()
+
+    def _remove_native_vlan(self, port):
+        """Removes native vlan from a trunked port.
+        If it is the last vlan to be removed, it disables the port and
+        reverts its state to default configuration
+        """
+        port_info = self._interface_info(port)
+        vlan_id = str(port_info[port]['native_vlan'])
+        self._set_commands(port, vlan_id)
+        remove_native_vlan = self.remove_native_vlan
+        if isinstance(port_info[port]['vlans'], (str, unicode)):
+            self.revert_port(port)
+        else:
+            self._set_commit_config(remove_native_vlan)
+        self._clear_set_commands()
+
+    def _get_mode(self, port):
+        """Returns True if the port is in trunk mode else returns False. """
```

Comment:
the name implies that it will return either trunk or access, not some boolean. Should choose a more appropriate name, something like `is_mode_trunk()`